applications/pusher/src/pusher_listener.erl

```diff
@@ -67,50 +67,71 @@ handle_push(JObj, _Props) ->
-                                              ,<<"Authorizing-ID">>
-                                              ], JObj),
-    maybe_update_push_token(AccountId, AuthorizingId, UA, JObj, Params).
+    case props:is_defined(TokenKey, Params) of
+        'false' ->
+            maybe_clear_push_token(AccountId, AuthorizingId, get_endpoint(AccountId, AuthorizingId));
+        'true' ->
+            maybe_update_push_token(AccountId, AuthorizingId
+                                   ,kz_json:set_value(<<"Token-Proxy">>, ?TOKEN_PROXY_KEY, UA)
+                                   ,JObj, Params, get_endpoint(AccountId, AuthorizingId)
+                                   )
+    end.
 
--spec maybe_update_push_token(kz_term:api_binary(), kz_term:api_binary(), kz_json:object(), kz_json:object(), kz_term:proplist()) -> 'ok'.
-maybe_update_push_token('undefined', _AuthorizingId, _UA, _JObj, _Params) -> 'ok';
-maybe_update_push_token(_AccountId, 'undefined', _UA, _JObj, _Params) -> 'ok';
-maybe_update_push_token(AccountId, AuthorizingId, UA, JObj, Params) ->
+-spec get_endpoint(kz_term:ne_binary(), kz_term:ne_binary()) -> kz_json:api_object().
+get_endpoint(AccountId, AuthorizingId) ->
     AccountDb = kz_util:format_account_db(AccountId),
     case kz_datamgr:open_cache_doc(AccountDb, AuthorizingId) of
-        {'ok', Doc} ->
-            Push = kz_json:get_value(<<"push">>, Doc),
-            case build_push(UA, JObj, Params, kz_json:new()) of
-                Push -> lager:debug("push exists: ~p", [Push]);
-                NewPush ->
-                    {'ok', _} = kz_datamgr:save_doc(AccountDb, kz_json:set_value(<<"push">>, NewPush, Doc)),
-                    lager:debug("setting push object for ~s: ~s: ~p", [AccountId, AuthorizingId, NewPush])
-            end;
-        {'error', _} -> lager:debug("failed to open ~s in ~s", [AuthorizingId, AccountId])
+        {'ok', Doc} -> Doc;
+        {'error', _} ->
+            lager:debug("failed to open ~s in ~s", [AuthorizingId, AccountId]),
+            'undefined'
+    end.
+
+-spec maybe_clear_push_token(kz_term:ne_binary(), kz_term:ne_binary(), kz_json:api_object()) -> 'ok'.
+maybe_clear_push_token(_, _, 'undefined') -> 'ok';
+maybe_clear_push_token(AccountId, AuthorizingId, Doc) ->
+    case kzd_devices:push(Doc) of
+        'undefined' -> 'ok';
+        _ ->
+            AccountDb = kz_util:format_account_db(AccountId),
+            {'ok', _} = kz_datamgr:save_doc(AccountDb, kzd_devices:delete_push(Doc)),
+            lager:debug("clearing push object for ~s: ~s", [AccountId, AuthorizingId])
+    end.
+
+-spec maybe_update_push_token(kz_term:ne_binary(), kz_term:ne_binary(), kz_json:object(), kz_json:object(), kz_term:proplist(), kz_json:api_object()) -> 'ok'.
+maybe_update_push_token(_, _, _, _, _, 'undefined') -> 'ok';
+maybe_update_push_token(AccountId, AuthorizingId, UA, JObj, Params, Doc) ->
+    Push = kz_json:get_value(<<"push">>, Doc),
```

Comment:
Use the accessor

Reply:
hey. I just copy-pasted that part 😄
I'll make sure to update it